README.md
# RCT2 Fleecer

RCT2 Fleecer is a small web app to help manage the entry fee and ride ticket prices for the games RollerCoaster Tycoon 2, RollerCoaster Tycoon Classic and OpenRCT2. Given a ride's type and ratings, it will calculate the maximum price that guests will pay for that ride. Build up a list of rides and see at a glance how much the rides in your park are worth!

[https://invalio19.github.io/rct2-fleecer/](https://invalio19.github.io/rct2-fleecer/)

## Features

- Build up a list of rides and see the maximum price you can charge for them (or the 'good value' price) in pay-per-ride scenarios.
- See the recommended park entrance fee to attract as many guests and make as much money as possible in pay-for-entry scenarios.
- Automatic duplicate ride type detection (having more than one ride of the same type decreases its value by about 25%).
- Data is auto-saved to the browser whenever any details are changed, and persists even when the page is closed.
- View a ride's minimum stat requirements, so you can build rides without rating penalties.
- Compatible with RCT2, RCT Classic and OpenRCT2.
- Manually sort the order of rides.
- Designed to work well on mobile.

## Planned features

See the [issue tracker](https://github.com/invalio19/rct2-fleecer/issues).

## Built with

- [Angular](https://angular.io/) v9.1.3.
- [Bulma](https://bulma.io/) v0.8.0.
- [Bulma Checkradio](https://wikiki.github.io/form/checkradio/) v1.1.1
- [Font Awesome](https://fontawesome.com/) v5.13.0

## Contributing

Feel free to fork this project and create your own pull requests.

Bugs and feature requests can be raised [here](https://github.com/invalio19/rct2-fleecer/issues). There are a number of planned/potential features here already so do check check that your request isn't already on the list!

## FAQs

**1. Why am I seeing the wrong ride price? Guests still refuse to go on my ride.**

There could be a few reasons for this.

- If you are charging a park entrance fee, make sure you've checked that checkbox since this will quarter the price you can charge.
- If you have more than one of the same type of ride in the park, this should be flagged automatically, and will reduce the price you can charge by about 25%.
- Have the ratings on your ride increased or decreased? Many things can change the ratings of a ride, including scenery and other nearby rides.
- Check that you have the right game version chosen of either 'RCT2 / RCT Classic' or 'OpenRCT2', as they have different calculation algorithms for the earlier months of a ride's life.
- If you genuinely think there is an error in the calculation, raise it as a bug [here](https://github.com/invalio19/rct2-fleecer/issues), providing as much information as you can give.

**2. Why are two of my rides considered duplicates of each other even though they are different ride types?**

Many rides in RCT2 are considered the same for some game mechanics, including the mechanics used for determining ride price value. These rides are:

- The **Car Ride** and **Monster Trucks**.
- The **Classic Mini Roller Coaster** (OpenRCT2 only) and **Junior Roller Coaster**.
- The **Corkscrew Roller Coaster** and **Hypercoaster**.
- The **Hyper-Twister** and **Steel Twister**.
- The **Spinning Wild Mouse** and **Steel Wild Mouse**.

There are also many more re-skinned versions of rides such as the **Fire Cracker Ride** simply being a copy of the **Enterprise**.

**3. My ride is six years old but the price calculation for 'less than 88 months old' is wrong. Why is that?**

A year in RollerCoaster Tycoon is only eight months long, from March to October, presumably to simulate when real-life parks are closed for the winter. You should select the number of months accordingly.

**4. I reduced the age of my coaster, yet its value went up. Why?**

After 200 months—25 years in RollerCoaster Tycoon time—the value of a ride actually goes up instead of down and will stay at that level forevermore. I would imagine this is to simulate the appeal of retro roller coasters in real life.

**5. How does this app save my data? Can I transfer it to another browser or device?**

RCT2 Fleecer uses LocalStorage to save data about the current page to your browser. Currently there is no in-app support for transferring this data to another browser or device.

**6. This app doesn't work on my browser. Why not?**

RCT2 Fleecer was largely built and tested using Google Chrome. Modern browsers like Firefox should also have no issues handling it, but older browsers such as Internet Explorer cannot.

## Changelog

See the [CHANGELOG](https://github.com/invalio19/rct2-fleecer/blob/master/CHANGELOG.md) file.

## Authors

- Shay Irwin: [Invalio](https://github.com/invalio19)

## License

- [MIT](http://opensource.org/licenses/mit-license.php). See the [LICENSE](https://github.com/invalio19/rct2-fleecer/blob/master/LICENSE) file for details.
- Copyright 2020 &copy; Invalio.

## Acknowledgements

- [Deurklink](https://www.youtube.com/channel/UCcU9si2fIVJ-KoIDX9xYpdw), for his [work](https://forums.openrct2.org/topic/2737-guide-how-much-can-you-charge-for-your-rides/) in uncovering how the ride pricing is determined.
- [Marcel Vos](https://www.youtube.com/channel/UCBlXovStrlQkVA2xJEROUNg), for [explaining](https://www.youtube.com/watch?v=rUGUwZIr4n0) how this and many other RCT2 mechanics work, as well as some very entertaining RCT2 videos.
- [shottysteve](https://www.youtube.com/shottysteve), for designing the [calculator](https://rct2calc.shottysteve.com/) that helped inspire this one.
- [Ted 'IntelOrca' John](http://intelorca.co.uk/) and the various contributors to the [OpenRCT2](https://openrct2.org/) project.

CHANGELOG.md
# Changelog

## [1.4.0](https://github.com/invalio19/rct2-fleecer/compare/v1.3.1...v1.4.0) - 2020-05-02

### Added

- All of the alternative ride type names (such as Fire Cracker Ride for the Enterprise) are now included in the ride type list.

### Changed

- Angular version updated to 9.1.3.
- Tweaked the layout to display full-width backgrounds for containers on desktop.
- Ride type drop-down list is now auto-sorted by name.

### Fixed

- This very changelog has the correct year for updates--we're no longer in 2019!

## [1.3.1](https://github.com/invalio19/rct2-fleecer/compare/v1.3.0...v1.3.1) - 2020-04-26

### Fixed

- Typo fix.

## [1.3.0](https://github.com/invalio19/rct2-fleecer/compare/v1.2.0...v1.3.0) - 2020-04-26

### Added

- There is now a choice between the three admission modes for a park, with prices calculated accordingly.

### Fixed

- The default ride name now has an incrementing number instead of always being '1'.

## [1.2.0](https://github.com/invalio19/rct2-fleecer/compare/v1.1.0...v1.2.0) - 2020-04-22

### Added

- Minimum stat requirements and corresponding rating penalties can now be viewed for rides that have them.
- The 'What's new?' modal appears for visitors who have seen the page on versions <= 1.1.0.

### Changed

- The 'upgrade ride age' button has been replaced with a 'refurbish' button which resets a ride's age.
- When the park has an entrance fee, the 'show good value prices' checkbox is hidden rather than disabled.
- Updated the checkbox styles.
- Updated the icons to Font Awesome 5.

### Fixed

- The ride price field and ride rating labels no longer act like clickable buttons.

## [1.1.0](https://github.com/invalio19/rct2-fleecer/compare/v1.0.1...v1.1.0) - 2020-04-17

### Added

- The recommended park entrance fee field is now calculated and displayed, along with a modal explaining its logic.

## [1.0.1](https://github.com/invalio19/rct2-fleecer/compare/v1.0.0...v1.0.1) - 2020-04-11

### Fixed

- 'Delete all rides' no longer prevents auto-saving from working until the page is refreshed.
- 'Delete all rides' no longer resets options such as game version to their default value.

## [1.0.0](https://github.com/invalio19/rct2-fleecer/releases/tag/v1.0.0) - 2020-04-09

### Added

- Build up a list of rides and get the maximum price you can charge for them—or the 'good value' price—based on its ratings.
- Automatic duplicate ride type detection (having more than one ride of the same type decreases its value by about 25%).
- Data is auto-saved to the browser whenever any details are changed, and persists even when the page is closed.
- Compatible with RCT2, RCT Classic and OpenRCT2.
- Manually sort the order of rides.
- Designed to work well on mobile.

src/app/data/ride-types.ts
import { RideType } from '../rides/shared/models/ride-type.model';

export const RideTypesData: { [key: string]: RideType } = {
  '3dCinema': {
    id: '3dCinema',
    name: '3D Cinema',
    excitement: 20,
    intensity: 10,
    nausea: 0
  },
  airPoweredVerticalCoaster: {
    id: 'airPoweredVerticalCoaster',
    name: 'Air Powered Vertical Coaster',
    excitement: 44,
    intensity: 66,
    nausea: 10
  },
  boatHire: {
    id: 'boatHire',
    name: 'Boat Hire',
    excitement: 70,
    intensity: 6,
    nausea: 0
  },
  bobsleighCoaster: {
    id: 'bobsleighCoaster',
    name: 'Bobsleigh Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  carRide: {
    id: 'carRide',
    name: 'Car Ride',
    excitement: 70,
    intensity: 10,
    nausea: 10
  },
  chairlift: {
    id: 'chairlift',
    name: 'Chairlift',
    excitement: 70,
    intensity: 10,
    nausea: 0
  },
  circus: {
    id: 'circus',
    name: 'Circus',
    excitement: 20,
    intensity: 10,
    nausea: 0
  },
  classicMiniRollerCoaster: {
    id: 'classicMiniRollerCoaster',
    name: 'Classic Mini Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  compactInvertedCoaster: {
    id: 'compactInvertedCoaster',
    name: 'Compact Inverted Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  corkscrewRollerCoaster: {
    id: 'corkscrewRollerCoaster',
    name: 'Corkscrew Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  crookedHouse: {
    id: 'crookedHouse',
    name: 'Crooked House',
    excitement: 15,
    intensity: 8,
    nausea: 0
  },
  dinghySlide: {
    id: 'dinghySlide',
    name: 'Dinghy Slide',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  dodgems: {
    id: 'dodgems',
    name: 'Dodgems',
    excitement: 40,
    intensity: 20,
    nausea: 0
  },
  enterprise: {
    id: 'enterprise',
    name: 'Enterprise',
    excitement: 50,
    intensity: 10,
    nausea: 0
  },
  ferrisWheel: {
    id: 'ferrisWheel',
    name: 'Ferris Wheel',
    excitement: 60,
    intensity: 20,
    nausea: 10
  },
  flyingRollerCoaster: {
    id: 'flyingRollerCoaster',
    name: 'Flying Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  flyingSaucers: {
    id: 'flyingSaucers',
    name: 'Flying Saucers',
    excitement: 50,
    intensity: 25,
    nausea: 0
  },
  ghostTrain: {
    id: 'ghostTrain',
    name: 'Ghost Train',
    excitement: 70,
    intensity: 10,
    nausea: 10
  },
  gigaCoaster: {
    id: 'gigaCoaster',
    name: 'Giga Coaster',
    excitement: 51,
    intensity: 32,
    nausea: 10
  },
  goKarts: {
    id: 'goKarts',
    name: 'Go-Karts',
    excitement: 120,
    intensity: 20,
    nausea: 0
  },
  hauntedHouse: {
    id: 'hauntedHouse',
    name: 'Haunted House',
    excitement: 20,
    intensity: 10,
    nausea: 0
  },
  heartlineTwisterCoaster: {
    id: 'heartlineTwisterCoaster',
    name: 'Heartline Twister Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  hyperTwister: {
    id: 'hyperTwister',
    name: 'Hyper-Twister',
    excitement: 52,
    intensity: 36,
    nausea: 10
  },
  hypercoaster: {
    id: 'hypercoaster',
    name: 'Hypercoaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  invertedHairpinCoaster: {
    id: 'invertedHairpinCoaster',
    name: 'Inverted Hairpin Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 30
  },
  invertedImpulseCoaster: {
    id: 'invertedImpulseCoaster',
    name: 'Inverted Impulse Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  invertedRollerCoaster: {
    id: 'invertedRollerCoaster',
    name: 'Inverted Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  juniorRollerCoaster: {
    id: 'juniorRollerCoaster',
    name: 'Junior Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  launchedFreefall: {
    id: 'launchedFreefall',
    name: 'Launched Freefall',
    excitement: 50,
    intensity: 50,
    nausea: 10
  },
  layDownRollerCoaster: {
    id: 'layDownRollerCoaster',
    name: 'Lay-down Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  lift: {
    id: 'lift',
    name: 'Lift',
    excitement: 80,
    intensity: 10,
    nausea: 0
  },
  limLaunchedRollerCoaster: {
    id: 'limLaunchedRollerCoaster',
    name: 'LIM Launched Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  logFlume: {
    id: 'logFlume',
    name: 'Log Flume',
    excitement: 80,
    intensity: 34,
    nausea: 6
  },
  loopingRollerCoaster: {
    id: 'loopingRollerCoaster',
    name: 'Looping Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  magicCarpet: {
    id: 'magicCarpet',
    name: 'Magic Carpet',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  maze: {
    id: 'maze',
    name: 'Maze',
    excitement: 50,
    intensity: 0,
    nausea: 0
  },
  merryGoRound: {
    id: 'merryGoRound',
    name: 'Merry-Go-Round',
    excitement: 50,
    intensity: 10,
    nausea: 0
  },
  mineRide: {
    id: 'mineRide',
    name: 'Mine Ride',
    excitement: 60,
    intensity: 20,
    nausea: 10
  },
  mineTrainCoaster: {
    id: 'mineTrainCoaster',
    name: 'Mine Train Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  miniGolf: {
    id: 'miniGolf',
    name: 'Mini Golf',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  miniHelicopters: {
    id: 'miniHelicopters',
    name: 'Mini Helicopters',
    excitement: 70,
    intensity: 10,
    nausea: 10
  },
  miniRollerCoaster: {
    id: 'miniRollerCoaster',
    name: 'Mini Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  miniSuspendedCoaster: {
    id: 'miniSuspendedCoaster',
    name: 'Mini Suspended Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  miniatureRailway: {
    id: 'miniatureRailway',
    name: 'Miniature Railway',
    excitement: 70,
    intensity: 6,
    nausea: -10
  },
  monorail: {
    id: 'monorail',
    name: 'Monorail',
    excitement: 70,
    intensity: 6,
    nausea: -10
  },
  monorailCycles: {
    id: 'monorailCycles',
    name: 'Monorail Cycles',
    excitement: 50,
    intensity: 10,
    nausea: 10
  },
  monsterTrucks: {
    id: 'monsterTrucks',
    name: 'Monster Trucks',
    excitement: 50,
    intensity: 10,
    nausea: 10
  },
  motionSimulator: {
    id: 'motionSimulator',
    name: 'Motion Simulator',
    excitement: 24,
    intensity: 20,
    nausea: 10
  },
  multiDimensionRollerCoaster: {
    id: 'multiDimensionRollerCoaster',
    name: 'Multi-Dimension Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  observationTower: {
    id: 'observationTower',
    name: 'Observation Tower',
    excitement: 80,
    intensity: 10,
    nausea: 0
  },
  reverseFreefallCoaster: {
    id: 'reverseFreefallCoaster',
    name: 'Reverse Freefall Coaster',
    excitement: 44,
    intensity: 66,
    nausea: 10
  },
  reverserRollerCoaster: {
    id: 'reverserRollerCoaster',
    name: 'Reverser Roller Coaster',
    excitement: 48,
    intensity: 28,
    nausea: 7
  },
  riverRafts: {
    id: 'riverRafts',
    name: 'River Rafts',
    excitement: 80,
    intensity: 34,
    nausea: 6
  },
  riverRapids: {
    id: 'riverRapids',
    name: 'River Rapids',
    excitement: 72,
    intensity: 26,
    nausea: 6
  },
  rotoDrop: {
    id: 'rotoDrop',
    name: 'Roto-Drop',
    excitement: 50,
    intensity: 50,
    nausea: 10
  },
  sideFrictionRollerCoaster: {
    id: 'sideFrictionRollerCoaster',
    name: 'Side-Friction Roller Coaster',
    excitement: 48,
    intensity: 28,
    nausea: 7
  },
  spaceRings: {
    id: 'spaceRings',
    name: 'Space Rings',
    excitement: 12,
    intensity: 4,
    nausea: 4
  },
  spinningWildMouse: {
    id: 'spinningWildMouse',
    name: 'Spinning Wild Mouse',
    excitement: 50,
    intensity: 30,
    nausea: 30
  },
  spiralRollerCoaster: {
    id: 'spiralRollerCoaster',
    name: 'Spiral Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  spiralSlide: {
    id: 'spiralSlide',
    name: 'Spiral Slide',
    excitement: 50,
    intensity: 10,
    nausea: 0
  },
  splashBoats: {
    id: 'splashBoats',
    name: 'Splash Boats',
    excitement: 80,
    intensity: 34,
    nausea: 6
  },
  standUpRollerCoaster: {
    id: 'standUpRollerCoaster',
    name: 'Stand-up Roller Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  steelTwister: {
    id: 'steelTwister',
    name: 'Steel Twister',
    excitement: 52,
    intensity: 36,
    nausea: 10
  },
  steelWildMouse: {
    id: 'steelWildMouse',
    name: 'Steel Wild Mouse',
    excitement: 50,
    intensity: 30,
    nausea: 30
  },
  steeplechase: {
    id: 'steeplechase',
    name: 'Steeplechase',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  submarineRide: {
    id: 'submarineRide',
    name: 'Submarine Ride',
    excitement: 70,
    intensity: 6,
    nausea: 0
  },
  suspendedMonorail: {
    id: 'suspendedMonorail',
    name: 'Suspended Monorail',
    excitement: 70,
    intensity: 6,
    nausea: -10
  },
  suspendedSwingingCoaster: {
    id: 'suspendedSwingingCoaster',
    name: 'Suspended Swinging Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  swingingInverterShip: {
    id: 'swingingInverterShip',
    name: 'Swinging Inverter Ship',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  swingingShip: {
    id: 'swingingShip',
    name: 'Swinging Ship',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  topSpin: {
    id: 'topSpin',
    name: 'Top Spin',
    excitement: 24,
    intensity: 20,
    nausea: 10
  },
  twist: {
    id: 'twist',
    name: 'Twist',
    excitement: 40,
    intensity: 20,
    nausea: 10
  },
  verticalDropRollerCoaster: {
    id: 'verticalDropRollerCoaster',
    name: 'Vertical Drop Roller Coaster',
    excitement: 52,
    intensity: 38,
    nausea: 10
  },
  virginiaReel: {
    id: 'virginiaReel',
    name: 'Virginia Reel',
    excitement: 30,
    intensity: 15,
    nausea: 25
  },
  waterCoaster: {
    id: 'waterCoaster',
    name: 'Water Coaster',
    excitement: 50,
    intensity: 30,
    nausea: 10
  },
  woodenRollerCoaster: {
    id: 'woodenRollerCoaster',
    name: 'Wooden Roller Coaster',
    excitement: 52,
    intensity: 33,
    nausea: 8
  },
  woodenWildMouse: {
    id: 'woodenWildMouse',
    name: 'Wooden Wild Mouse',
    excitement: 50,
    intensity: 30,
    nausea: 30
  },
};